app/scripts/lib/rpc-method-middleware/handlers/wallet-upgrade-account.ts

```diff
@@ -0,0 +1,289 @@
+};
+
+export type UpgradeAccountResult = {
+  transactionHash: string; // Hash of the EIP-7702 authorization transaction
+  upgradedAccount: string; // Address of the upgraded account (same as input)
+  delegatedTo: string; // Address of the contract delegated to (determined by wallet)
+};
+
+export type GetAccountUpgradeStatusParams = {
+  account: string; // Address of the account to check
+  chainId?: number; // Optional: chain ID for the check (defaults to current)
+};
+
+export type GetAccountUpgradeStatusResult = {
+  account: string; // Address of the checked account
+  isUpgraded: boolean; // Whether the account is upgraded
+  chainId: number; // Chain ID where the check was performed
+};
+
+export const upgradeAccountHandler = {
+  methodNames: ['wallet_upgradeAccount'],
+  implementation: upgradeAccountImplementation,
+  hookNames: {
+    upgradeAccount: true,
+    getCurrentChainId: true,
+    isAtomicBatchSupported: true,
+  },
+};
+
+export const getAccountUpgradeStatusHandler = {
+  methodNames: ['wallet_getAccountUpgradeStatus'],
+  implementation: getAccountUpgradeStatusImplementation,
+  hookNames: {
+    getCurrentChainId: true,
+    getCode: true,
+    getNetworkConfigurationByChainId: true,
+  },
+};
+
+async function upgradeAccountImplementation(
+  req: JsonRpcRequest<Json[]> & { origin: string },
+  res: PendingJsonRpcResponse<Json>,
+  _next: AsyncJsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  {
+    upgradeAccount,
+    getCurrentChainId,
+    isAtomicBatchSupported,
+  }: {
+    upgradeAccount: (
+      address: string,
+      upgradeContractAddress: string,
+      chainId?: number,
+    ) => Promise<{ transactionHash: string; delegatedTo: string }>;
+    getCurrentChainId: () => number;
+    isAtomicBatchSupported: (request: {
+      address: string;
+      chainIds: string[];
+    }) => Promise<
+      {
+        chainId: string;
+        isSupported: boolean;
+        delegationAddress?: string;
+        upgradeContractAddress?: string;
+      }[]
+    >;
+  },
+) {
+  const { params, origin } = req;
+
+  // Check if the origin is a preinstalled snap. If its flask we allow any origin for local development.
+  if (!isFlask() && !isSnapPreinstalled(origin as SnapId)) {
+    return end(
+      rpcErrors.methodNotFound({
+        message:
+          'wallet_upgradeAccount is only available to preinstalled snaps',
+      }),
+    );
+  }
+
+  if (!Array.isArray(params) || params.length === 0) {
+    return end(
+      rpcErrors.invalidParams({
+        message: 'Expected non-empty array parameter',
+      }),
+    );
+  }
+
+  const [upgradeParams] = params as unknown as [UpgradeAccountParams];
+  const { account, chainId } = upgradeParams;
+
+  if (!account) {
+    return end(
+      rpcErrors.invalidParams({
+        message: 'account address is required',
+      }),
+    );
+  }
+
+  // Use current chain ID if not provided
+  const targetChainId = chainId ?? getCurrentChainId();
```

Comment:
I left a comment on the hook implementation too - as I understand it there is now an individual current chain per origin, so can we just call `getCurrentChainIdForDomain(origin)` to ensure that the currently selected chain for the specific caller is used?

(take this with salt, because: I don't know what the NetworkController state property signifies)